### ISSUE TITLE
omitNothingFields for Option

### ISSUE THREAD

**Issue by lspitzner** [bos/aeson#498]
Would it be possible to extend the behaviour of `omitNothingFields` to the `Data.Semigroups.Option` newtype `Maybe`-wrapper? "omit(Option)NothingFields"

**Comment by lspitzner** [comment]
also, with
~~~~.hs
data Foo t = Foo { x :: t Int, y :: t Bool }
~~~~
`omitNothingFields` does not seem to apply to the instances of `Foo Maybe`. I meant to confirm that `Foo Maybe` works where `Foo Option` did not, but it failed as well. It is possible i named the wrong issue in the title (Maybe/Option being used indirectly, via type argument, not Maybe vs. Option).

**Comment by lspitzner** [comment]
Ok, apparently `Maybe` works (even for the example `Foo`), but only while decoding; the result of encoding still contains "null"s for `Nothing`s.

Maybe someone could confirm this observation? If true, it seems worthy of a separate issue..

**Comment by phadej** [comment]
... and I won't be surprised if TH and Generics derivation differs here.

**Comment by lspitzner** [comment]
I did my testing with the Generics derivation.

**Comment by Lysxia** [comment]
I don't have any issue using `Generics`.

```haskell
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE OverloadedLists #-}

import Control.Monad
import Data.Maybe
import Data.Aeson
import Data.Aeson.Types
import GHC.Generics
    
data Foo t = Foo { x :: t Int, y :: t Bool } deriving Generic
    
main = do
  let opts  = defaultOptions{omitNothingFields=True}
      json  = genericToJSON opts (Foo Nothing (Just True))
      parse = genericParseJSON opts :: Value -> Parser (Foo Maybe)
  when (json /= Object [("y", Bool True)]) $ fail "encode"
  when (isNothing $ parseMaybe parse json) $ fail "decode"
```

**Comment by lspitzner** [comment]
right. It is likely that my diagnosis was flawed due to the misunderstanding I described in #518. The main topic of this issue seems to remain valid, however.

**Comment by Lysxia** [comment]
Oh I see, sorry. As for the actual issue, I think that is a good idea.

**Comment by lspitzner** [comment]
nah, i am sorry for the false alarm :)

**Comment by bergmark** [comment]
Sorry for not weighing in earlier. This sounds like a good change to me!

**Comment by bergmark** [comment]
This was resolved by #583 and is in aeson 1.2.3.0!
